test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java

```diff
@@ -896,6 +814,268 @@ public void sendNotification() {
+        setupExternalIdTest(false, true);
+        when(lookup.getString(EXTERNAL_IDENTIFIER)).thenReturn(EXTERNAL_ID);
+        when(optionsService.getOptions(HEALTH_CODE)).thenReturn(lookup);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NO_ID_PARTICIPANT);
+        
+        verifyNotSetAsReservation();
+        verifySetAsOption(null);
+    }
+
+    @Test(expected = BadRequestException.class)
+    public void updateExternalIdValidatedNotRequiredNoValue() {
+        setupExternalIdTest(true, false);
+        when(lookup.getString(EXTERNAL_IDENTIFIER)).thenReturn(EXTERNAL_ID);
+        when(optionsService.getOptions(HEALTH_CODE)).thenReturn(lookup);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NO_ID_PARTICIPANT);
+    }
+    
+    @Test
+    public void updateExternalIdNotValidatedNotRequiredNoValue() {
+        setupExternalIdTest(false, false);
+        when(lookup.getString(EXTERNAL_IDENTIFIER)).thenReturn(EXTERNAL_ID);
+        when(optionsService.getOptions(HEALTH_CODE)).thenReturn(lookup);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NO_ID_PARTICIPANT);
+        
+        verifyNotSetAsReservation();
+        verifySetAsOption(null);
+    }
+    
+    // Updating the participant with an external ID when one was not provided on creation
+    // (even if that would have been prevented by the code, just in case)
+    
+    // It's after creation so we just don't do it... of all the tests, this one is 
+    // debatable. If the value is supplied and its correct, even though it was required
+    // at sign up and should be there, we should probably add it now.
+    @Test
+    public void updateExternalIdValidatedRequiredNewValue() {
+        setupExternalIdTest(true, true);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NEW_ID_PARTICIPANT);
+        
+        verifyNotSetAsReservation();
+        verifyNotSetAsOption();
+    }
+
+    @Test
+    public void updateExternalIdNotValidatedRequiredNewValue() {
+        setupExternalIdTest(false, true);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NEW_ID_PARTICIPANT);
+        
+        verifyNotSetAsReservation();
+        verifySetAsOption("newExternalId");
+    }
+
+    @Test
+    public void updateExternalIdValidatedNotRequiredNewValue() {
+        setupExternalIdTest(true, false);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NEW_ID_PARTICIPANT);
+        
+        verifyIdAssignedWithoutReservation("newExternalId");
+        verifyNotSetAsOption();
+    }
+    
+    @Test
+    public void updateExternalIdNotValidatedNotRequiredNewValue() {
+        setupExternalIdTest(false, false);
+        
+        participantService.updateParticipant(STUDY, CALLER_ROLES, NEW_ID_PARTICIPANT);
+        
+        verifyNotSetAsReservation();
+        verifySetAsOption("newExternalId");
+    }    
```

Comment:
I noticed that you've thoroughly exhaustively tested every possibility. Generally speaking, this is a very commendable thing.

However, this also tells me that saveParticipant() is growing too complex. This is not sustainable, as its unit tests will continue to grow exponentially, and eventually we won't be able change it at all.

I don't know if now is the right time to do this, but at some point before the next update to saveParticipants(), we should refactor the code so that the complexity and test requirements of this class grow linearly rather than exponentially. At the very least, if we can isolate some of the variables from one another, we can reduce some of the complexity and avoid having to write 2^N tests.